proposals/79-typescript-definitions.md

```diff
@@ -0,0 +1,54 @@
+# TypeScript Definitions for core libraries
+
+Authors: @rvagg
+
+Initial PR: https://github.com/protocol/web3-dev-team/pull/79
+
+## Purpose &amp; impact 
+
+Consumers of our JavaScript libraries & components should have sufficient TypeScript definitions, available through standard means, to write fully typed TypeScript code, or use that code to drive tooling that consumes these definitions - such as VS Code editing enhancements, documentation production pipelines, type checking tooling for test & CI pipelines.
+
+This work is high-value and high-impact for the JavaScript ecosystem and those of us working on open source JavaScript libraries can all provide anecdotal evidence for the frequency with which developers request better TypeScript annotations for our libraries. The rate of TypeScript adoption, particularly within larger-scale projects, is increasing, but TypeScript definitions are increasingly useful as they are included in general JavaScript linting, checking and documentation tooling.
+
+This work provides high-leverage within our suite of JavaScript tools as we mature, refactor, modularize and create. We are already establishing a suite of practices and tooling that are used in varying ways across PL JavaScript projects that use definitions, even though we have very few TypeScript projects throughout our GitHub orgs (<https://github.com/ipfs/js-dag-service> being a rare example, which was initially contributed by Textile). It is reasonable to expect that the majority of new JavaScript code produced by Protocol Labs into the future will make use of TypeScript annotations in some way.
+
+Work on this effort has been largely completed, thanks primarily to @hugomrdias, @Gozala, and @achingbrain with assistance from the entire JavaScript team in various ways. The bulk of the activity has been tracked as a roll-up in to js-ipfs @ <https://github.com/ipfs/js-ipfs/issues/2945>.
+
+Remaining work to integrate the current set of js-ipfs dependencies into js-ipfs and address the typing for code directly in js-ipfs is happening here: https://github.com/ipfs/js-ipfs/pull/3550
+
+Aside from completing the remaining js-ipfs itegration work, the scope of this project includes some additional libraries that re not currently part of the js-ipfs dependency tree, including:
+
+ * Next-generation IPLD codec libraries (using the js-multiformats pattern)
```

Comment:
Does this include multiaddr update? https://github.com/multiformats/js-multiaddr/pull/159

Reply:
I think it probably should be, but I don't have a feel for how much work it is to get that over the line. Do you?

Reply:
maybe @hugomrdias have an idea on the current state. More than having that merged, this will require changes across the stack as this will change how multiaddrs are created

Reply:
That PR should be made non-breaking just to get it finished, i tried a week ago or something but its troublesome. Need to look at it again.